pkg/manifests/manifests.go

```diff
@@ -772,6 +774,94 @@ func (f *Factory) KubeStateMetricsDeployment() (*appsv1.Deployment, error) {
 	return d, nil
 }
 
+// KubeStateMetricsDenylistBoundsCheck ensures that the user is:
+// * able to enable metrics that are denied by default, and,
+// * unable to disable metrics that are enabled by default.
+func (f *Factory) KubeStateMetricsDenylistBoundsCheck(deployment *appsv1.Deployment, service *v1.Service, secret *v1.Secret) (*appsv1.Deployment, error) {
```

Comment:
Clever idea :) but I have some concerns:

- Making an operator depend on one of its operands doesn't seem that common and natural to me. Here, for example, if CMO cannot reach out to KSM's` /metrics`, it'll abort its task and not try to "help it". This could be fixed, but maybe there will be other deadlocks like this.
- Does KSM always expose all the metrics that it can expose? I mean, does it initialize all the metrics? If not, `/metrics` will not really be a source of truth: A user could drop a not-there-yet metric.
- The code seems a little bit hacky as we're re-implementing some parts of a scraper, and I'm concerned about its maintainability.

I don't have a better alternative but I see that for the majority of metrics in the current `--metric-denylist` we specify the exact metrics. How about expanding the ones with `xxx_.+_yyy` and just allowing users to select from that list? But maybe it'll be harder to maintain. (we can still have tests with `xxx_.+_yyy` to be sure we take new metrics we want to drop into account)

Maybe there is also something to be done in connection with collection profiles...

Reply:
> Making an operator depend on one of its operands doesn't seem that common and natural to me. Here, for example, if CMO cannot reach out to KSM's /metrics, it'll abort its task and not try to "help it". This could be fixed, but maybe there will be other deadlocks like this.

Ah, I believe CMO itself won't lean on KSM anymore than it did before this patch. A bounds-check for an intrinsic property (`--metric-denylist`) of the operand is nothing more than its own behavior being handled by CMO, same as any other non-repairable (by choice or by trial) error it'd exhibit (for instance, an invalid CRS config). We are failing KSM reconciliation and throwing that error without making any assumptions or implicit conversions to be explicit towards the user on CMO's resolution of their supplied deny-list. PLMK if I missed something.

> Does KSM always expose all the metrics that it can expose? I mean, does it initialize all the metrics? If not, /metrics will not really be a source of truth: A user could drop a not-there-yet metric.

Since help-texts will always be present in the exposition data, so [we should be safe](https://github.com/openshift/cluster-monitoring-operator/pull/2283/files#diff-1f9d2d18d4cabd8262e585b932330183f7cb9b544ac9f27c8bcb3ef9c8d14376R840-R841).

> [...] re-implementing some parts of a scraper [...]

Do we already have helpers for this in CMO? The single-scrape logic essentially crafts a request and interprets it, while pinging the `/metrics` endpoint with the necessary auth permissions. It seemed like a straightforward implementation, but feel free to drop any suggestions (in addition to Jan's comment above, as I've made those changes locally but I want to solve the auth issue before I push again) that may abstract this and reduce maintainability costs.

Reply:
> Ah, I believe CMO itself won't lean on KSM anymore than it did before this patch. A bounds-check for an intrinsic property (--metric-denylist) of the operand is nothing more than its own behavior being handled by CMO, same as any other non-repairable (by choice or by trial) error it'd exhibit (for instance, an invalid CRS config). We are failing KSM reconciliation and throwing that error without making any assumptions or implicit conversions to be explicit towards the user on CMO's resolution of their supplied deny-list. PLMK if I missed something.

I see, but in this case, it seems to me that if KSM itself is broken (or CMO thinks it’s broken), CMO will not sync KSM, which isn’t the behavior we expect from an operator. For example, if one makes an unintentional change to the KSM Deployment that broke it, CMO will abandon it instead of fixing it as expected.
CMO being unable to proceed because of some third-party dependency that CMO cannot even control/adjust (CR, a resource managed by another operator e.g.) is okay, but CMO not being able to fix an operand because the broken operand isn’t responding to CMO, well, that’s a circular dependency :)
Again, this is just an example, maybe there are others.

> Since help-texts will always be present in the exposition data, so [we should be safe](https://github.com/openshift/cluster-monitoring-operator/pull/2283/files#diff-1f9d2d18d4cabd8262e585b932330183f7cb9b544ac9f27c8bcb3ef9c8d14376R840-R841).

Ok, then KSM initializes all the metrics, (even the empty ones), good to know.

> Do we already have helpers for this in CMO? The single-scrape logic essentially crafts a request and interprets it, while pinging the /metrics endpoint with the necessary auth permissions. It seemed like a straightforward implementation, but feel free to drop any suggestions (in addition to Jan's comment above, as I've made those changes locally but I want to solve the auth issue before I push again) that may abstract this and reduce maintainability costs.

I don't think there are any existing utils for this (the only similar ones I see are used for tests), it'll be a new "trait/responsibility" that I'm not sure we should be adding to CMO. But the alternative I'm proposing (to expand the`xxx_.+_yyy`) isn't without its flaws.
Of course, if we agree to merge this, maybe we should consider moving that new logic from `manifests.go` and adding some tests.

Reply:
IMHO this boils down to a question of trading-off between:

* failing on an invalid deny-list, albeit at the expense of degrading a healthy KSM instance, or,
* subtly emitting or logging the issue, and relying on the editor to diagnose the issue.

I guess I'm okay with wherever the team lands on this.

> Ok, then KSM initializes all the metrics, (even the empty ones), good to know.

Ah, sorry, I believe I should've been more explicit. So KSM has a [list of default resources](https://github.com/kubernetes/kube-state-metrics/blob/93e79fc9d8741c5e605fdab0e56700fdc320c7f1/pkg/options/resource.go#L27) that it builds metric stores for from the get-go, but these have to be present in the cluster in order for the informers to trigger the metric family population. If so, these are initialized, and I believe this set should be a safe assumption for all OpenShift variants (I've tested this on `hypershift`, but not all of them individually). LMK if you think otherwise.

Reply:
Thanks Ayoub for raising this, I think its well worth thinking carefully about the failure scenarios here.
Afaiu Ayoubs concern is that an erroneous user supplied deny list will cause unrelated tasks to fail. In the current iteration this is true. If `KubeStateMetricsDenylistBoundsCheck` returns an error the KSM PrometheusRule and ServiceMonitor will not be reconciled. Also the config sharing task (in its own task group) will not run. 

Maybe the application of the user supplied deny list should run in its own task in the last task group. If that fails at least the remaining stack is already rolled out completely.

Reply:
Thanks for this rubberducking session :)

Actually, I thought that `CreateOrUpdateDeployment` call was moved after `KubeStateMetricsDenylistBoundsCheck`. That's why I was talking about a deadlock: "If CMO cannot reachKSM /metrics because a user broke the `Deployment` resource on the cluster with a non-denylist related change, CMO will not update the `Deployment` to try to fix KSM...". However, in truth, we call `CreateOrUpdateDeployment` both before and after. My bad.

Still, as Jan mentioned, `KubeStateMetricsDenylistBoundsCheck` may prevent the following `PrometheusRule/ServiceMonitor` related operations. But if `KubeStateMetricsDenylistBoundsCheck` **can only** return an error in case of an erroneous user-supplied denylist, I think it's okay to fail the task. I don't believe we validate at 100% all user inputs elsewhere. (If a user really wants to break a component, they'd be able to do so).

If the "expand the `xxx_.+_yyy` metrics and let them choose from them" approach isn't appealing, why couldn't we just pass the input as is? As we do for the net interfaces, e.g.: https://github.com/openshift/cluster-monitoring-operator/blob/4b8086df25735168099a7f18ab78deaa30b5c73d/pkg/manifests/config.go#L306-L320. I'm not sure, but maybe excluding the wrong interface can already "break" some stuff.

Besides the ones for alerts, dashboards, etc., which metrics can we not afford to lose? Couldn't we define them explicitly as part of a "minimal collection profile" and ensure they're never excluded? (This slightly aligns with the "expand the `xxx_.+_yyy` metrics" approach). This way, we only have one method for defining required/minimal metrics: collection profiles.

(Sorry, if this was discussed elsewhere, please point me to that)

Reply:
> However, in truth, we call CreateOrUpdateDeployment both before and after. My bad.

I'm not sure I understand. KSM's task will still error out if a non-empty invalid deny-list is supplied.

> Besides the ones for alerts, dashboards, etc., which metrics can we not afford to lose? Couldn't we define them explicitly as part of a "minimal collection profile" and ensure they're never excluded? (This slightly aligns with the "expand the xxx_.+_yyy metrics" approach). This way, we only have one method for defining required/minimal metrics: collection profiles.

The problem with doing that, is multi-staged and multi-faceted. Note that some bullets below may go outside of the scope of your question, but I've included them for the sake of arriving on a consensus.

- We cannot exhaustively _lasso_ all metrics that critical OpenShift components depend upon at **any** given time, unless we have [a set of `targetSelector`s](https://github.com/rexagod/cpv/tree/main#extraction) that allows us to do that.
- [From the point-of-view of the users being able to customize denied metrics, and CPs being a deny-list alternative] If CMO was to "inject" the user's deny-list regex expressions into the profile (or give them any control over it), this will allow users to "break" CMO's "minimal" commitment, rendering the full and minimal profiles similar, etc.
- CPs that the CMO is responsible for are not meant to be dynamic (as in, configurable-on-the-fly). Using the deny-list option in the configuration, users have reasonable control over their **generated metrics**, without CMO reconciling them back. Additionally, CPs will choke out a lot of high-cardinality metrics **after they have been generated**, rather than not generating them at all (deny-list).
- We don't have a concept of per-component collection profiles yet, but hypothetically, if we had a `custom` collection profile for custom use cases (disabling or enabling a set of metrics that don't fit into the `minimal` profile's characteristics), that will **disallow users to apply other profiles** at the same time. This means that if they want a set of metrics denied, but at the same time have a **cluster-wide** `minimal` profile (or any other profile that's different from the currently applied (`custom`) profile that has denied metrics at the user's discretion) that they want to apply, they won't be able to do so. We will lose the ability to decouple (a) "denying" a set of metrics from (b) "defining" a set of metrics that establish various cluster behaviors.
- `drop`ing metrics in the KSM `minimal` service monitor will replace a KSM behavior, that it already has explicit support for, with a CMO responsibility entirely. IMHO choosing between extending an already existing behavior that a component exhibits (as we do stack-wide for a lot of them) versus not using that at all and replacing that with a CMO-specific trait is not ideal, given that we already rely on deny-listing a set of default metrics in the KSM deployment. It only makes sense to extend it as compared to hooking that to one of KSM's profiles (`minimal`, `custom`, etc.).

> But if KubeStateMetricsDenylistBoundsCheck can only return an error in case of an erroneous user-supplied denylist, I think it's okay to fail the task.

IIUYC we'd still need to modify the deployment after the bounds check, to make the valid user deny-list come into effect.

> [...] why couldn't we just pass the input as is?

Regexes are supported in the `kubeStateMetrics.metricDenylist`, if that's what you meant here.

***

> Maybe the application of the user supplied deny list should run in its own task in the last task group. If that fails at least the remaining stack is already rolled out completely.

ACK, I'll send a patch.

Reply:
> Besides the ones for alerts, dashboards, etc., which metrics can we not afford to lose? Couldn't we define them explicitly as part of a "minimal collection profile" and ensure they're never excluded? (This slightly aligns with the "expand the xxx_.+_yyy metrics" approach). This way, we only have one method for defining required/minimal metrics: collection profiles.

The problem with doing that, is multi-staged and multi-faceted. Note that some bullets below may go outside of the scope of your question, but I've included them for the sake of arriving on a consensus.

- We cannot exhaustively _lasso_ all metrics that critical OpenShift components depend upon at **any** given time, unless we have [a set of `targetSelector`s](https://github.com/rexagod/cpv/tree/main#extraction) that allows us to do that.
- [From the point-of-view of the users being able to customize denied metrics, and CPs being a deny-list alternative] If CMO was to "inject" the user's deny-list regex expressions into the profile (or give them any control over it), this will allow users to "break" CMO's "minimal" commitment, rendering the full and minimal profiles similar, etc.
- CPs that the CMO is responsible for are not meant to be dynamic (as in, configurable-on-the-fly). Using the deny-list option in the configuration, users have reasonable control over their **generated metrics**, without CMO reconciling them back. Additionally, CPs will choke out a lot of high-cardinality metrics **after they have been generated**, rather than not generating them at all (deny-list).
- We don't have a concept of per-component collection profiles yet, but hypothetically, if we had a `custom` collection profile for custom use cases (disabling or enabling a set of metrics that don't fit into the `minimal` profile's characteristics), that will **disallow users to apply other profiles** at the same time. This means that if they want a set of metrics denied, but at the same time have a **cluster-wide** `minimal` profile (or any other profile that's different from the currently applied (`custom`) profile that has denied metrics at the user's discretion) that they want to apply, they won't be able to do so. We will lose the ability to decouple (a) "denying" a set of metrics from (b) "defining" a set of metrics that establish various cluster behaviors.
- `drop`ing metrics in the KSM `minimal` service monitor will replace a KSM behavior, that it already has explicit support for, with a CMO responsibility entirely. IMHO choosing between extending an already existing behavior that a component exhibits (as we do stack-wide for a lot of them) versus not using that at all and replacing that with a CMO-specific trait is not ideal, given that we already rely on deny-listing a set of default metrics in the KSM deployment. It only makes sense to extend it as compared to hooking that to one of KSM's profiles (`minimal`, `custom`, etc.).

> But if KubeStateMetricsDenylistBoundsCheck can only return an error in case of an erroneous user-supplied denylist, I think it's okay to fail the task.

IIUYC we'd still need to modify the deployment after the bounds check, to make the valid user deny-list come into effect.

> [...] why couldn't we just pass the input as is?

Regexes are supported in the `kubeStateMetrics.metricDenylist`, if that's what you meant here.

***

> Maybe the application of the user supplied deny list should run in its own task in the last task group. If that fails at least the remaining stack is already rolled out completely.

ACK, I'll send a patch.

Reply:
Thanks for sharing that.
I wasn't suggesting dynamically changing CPs, but using the minimal CP as the source of truth: make `KubeStateMetricsDenylistBoundsCheck` rely on it instead of the `/metrics` output. If we had a minimal CP for KSM, CMO could only accept `kubeStateMetrics.metricDenylist` that doesn't drop metrics from the minimal CP.
Again, I'm just thinking out loud, maybe this needs more discussion later or somewhere else.

> Regexes are supported in the kubeStateMetrics.metricDenylist, if that's what you meant here.

Actually, I was suggesting to pass the user input to KSM as is, without any validation in CMO, like we do for the node-exporter net interfaces I shared above.

> ACK, thank you for clearing this up. I'll send a patch that should establish common ground.

Maybe we can just run that **at the end** of the KSM task (if it's possible)? That way, we're sure it's not blocking anything. This will save us the overhead of adding another task and all the conflicts and races that may occur if the two tasks run concurrently.

Reply:
> I wasn't suggesting dynamically changing CPs, but using the minimal CP as the source of truth: make KubeStateMetricsDenylistBoundsCheck rely on it instead of the /metrics output. If we had a minimal CP for KSM, CMO could only accept kubeStateMetrics.metricDenylist that doesn't drop metrics from the minimal CP.
Again, I'm just thinking out loud, maybe this needs more discussion later or somewhere else.

This is definitely worth exploring. I'd suggest through we wait to collection profiles are GA to pursue this.

> Maybe we can just run that at the end of the KSM task (if it's possible)? That way, we're sure it's not blocking anything. This will save us the overhead of adding another task and all the conflicts and races that may occur if the two tasks run concurrently.

This would still mean that the last task group wouldn't run (config sharing task). I think running it in the last task group is the best bang for the buck.

Reply:
It seems I missed the updates on this thread.

> I wasn't suggesting dynamically changing CPs, but using the minimal CP as the source of truth: make KubeStateMetricsDenylistBoundsCheck rely on it instead of the /metrics output. If we had a minimal CP for KSM, CMO could only accept kubeStateMetrics.metricDenylist that doesn't drop metrics from the minimal CP.
Again, I'm just thinking out loud, maybe this needs more discussion later or somewhere else.

That is doable if the metric expressions within KSM's minimal service monitor are not regexes themselves (at the moment, these are deterministic values surrounded by `|`, which can be used to statically validate the user-defined list). However,
- if they are, we'll need to either (a) switch to minimal profile every time we want to validate the denylist (to compile the regex, same problem we solve here by querying `/metrics`), or (b) have a complex regex implementation that verifies if a particular regex is a subset of the other, and additionally,
- this will also significantly increase the number of metrics that can be denied by the user in a non-minimal environment, which other components may depend upon (the number of metrics in KSM's `minimal` SM `keep` is much lesser than the metrics not in the default deny-list, giving the user much more control).

I guess this boils down to the fact that making KSM's `minimal` CP the source of truth for all non-`minimal` CP environments can be guaranteed to safeguard all metrics that will ever be relied on by any internal component. Note that while CP may be moving to GA, components will still need time to adapt to the `minimal` environment as there is no such concept being enforced currently (this is exactly what [CPV](https://github.com/rexagod/cpv) aims to help developers with).

> Actually, I was suggesting to pass the user input to KSM as is, without any validation in CMO, like we do for the node-exporter net interfaces I shared above.

By "validation", I'm assuming you mean the default-denylist subset-check that we do here, in which case, not doing so would give the user way too much control and break the "flexibility-stability" bounds this PR rests on. Not sure I follow the idea here, but feel free to correct me in case I misunderstood.

Reply:
Friendly ping. @machine424 PLMK if you'd prefer for me to schedule a call to discuss this thread, since reaching a consensus here in time may help us get this in 4.16. I deferred scheduling a call till now as I thought the discussion may end here and not need a call at all, but PLMK if you think otherwise, and if that would help accelerate things. Thanks!

Reply:
Thanks Pranshu for your answer.
(I don't think we need to switch to `minimal`, as CMO always knows about the SM manifest as it's in its assets/, we could also use the current profile instead of the minimal one)
As I mentioned, I'm not trying/aiming to block this PR, even though I don’t 100% agree with the approach ;) If the other colleague are okay with it, let's merge it.
Let's resume the profiles related discussion once they're GA or during their transition to GA.

Reply:
> I don't think we need to switch to minimal, as CMO always knows about the SM manifest as it's in its assets/, we could also use the current profile instead of the minimal one.

It's not possible to statically compile regexes (PTAL at my first point [above](https://github.com/openshift/cluster-monitoring-operator/pull/2283#discussion_r1586906524)), or atleast without a lot of computational (and maintenance) complexity.

***

> As I mentioned, I'm not trying/aiming to block this PR, even though I don’t 100% agree with the approach ;) If the other colleague are okay with it, let's merge it.

I'm not sure I see the cause for the incorporation of collection profiles here at the moment. Nonetheless, your review has been constructive and critical to this patch, and not at all blocking IMHO. I'd much rather defer this PR to after 4.16, than rush it right now against your acumen. Besides, we have little to gain from merging this PR in 4.16, as there are no such commitments that would tie us to doing so.

Reply:
I'll setup a call later this week to continue this discussion. That should make it easier to arrive at a consensus. :)